relay-server/src/services/processor.rs

```diff
@@ -651,38 +677,80 @@ impl ProcessingExtractedMetrics {
 
     #[cfg(feature = "processing")]
     fn retain(&mut self, mut f: impl FnMut(&Bucket) -> bool) {
-        self.0.project_metrics.retain(&mut f);
-        self.0.sampling_metrics.retain(&mut f);
+        self.metrics.project_metrics.retain(&mut f);
+        self.metrics.sampling_metrics.retain(&mut f);
     }
 
-    fn send_metrics(self, envelope: &Envelope, project_cache: Addr<ProjectCache>) {
-        let project_key = envelope.meta().public_key();
+    fn extrapolate(&self, buckets: &mut [Bucket]) {
+        let Some(factor) = self.extrapolation_factor else {
+            return;
+        };
 
-        let ExtractedMetrics {
-            project_metrics,
-            sampling_metrics,
-        } = self.0;
+        let extrapolate = match self.project.config().metric_extraction {
+            ErrorBoundary::Ok(ref config) if !config.extrapolate.is_empty() => &config.extrapolate,
+            _ => return,
+        };
 
-        if !project_metrics.is_empty() {
-            project_cache.send(AddMetricBuckets::internal(project_key, project_metrics));
-        }
+        let duplication = (factor.to_f64().round() as usize)
+            .min(self.global.options.extrapolation_duplication_limit)
+            .max(1);
 
-        if !sampling_metrics.is_empty() {
-            // If no sampling project state is available, we associate the sampling
-            // metrics with the current project.
-            //
-            // project_without_tracing         -> metrics goes to self
-            // dependent_project_with_tracing  -> metrics goes to root
-            // root_project_with_tracing       -> metrics goes to root == self
-            let sampling_project_key = utils::get_sampling_key(envelope).unwrap_or(project_key);
-            project_cache.send(AddMetricBuckets::internal(
-                sampling_project_key,
-                sampling_metrics,
-            ));
+        for bucket in buckets {
+            if !extrapolate.matches(&bucket.name) {
+                continue;
+            }
+
+            match bucket.value {
+                BucketValue::Counter(ref mut counter) => {
+                    *counter = counter.saturating_mul(factor);
+                }
+                BucketValue::Distribution(ref mut dist) => {
+                    *dist = std::mem::take(dist)
+                        .into_iter()
+                        .flat_map(|f| std::iter::repeat(f).take(duplication))
```

Comment:
Would be good to add a code comment somewhere that mentions that we want to move the repetition to the kafka producer (or close to the kafka producer) to save memory and minimize the rounding error.

Ideally Relay supports weighted buckets in the full pipeline, and only converts the weights in the last moment, so that it is easy to switch once storage supports weights.